lib/winapi/timezoneapi.c

```diff
@@ -0,0 +1,197 @@
+
+    // Check if daylight savings time (DST) adjustment is enabled
+    queryStatus = ExQueryNonVolatileSetting(XC_MISC, &type, &dayLightEnFlag, sizeof(dayLightEnFlag), NULL);
+    if (!NT_SUCCESS(queryStatus)) {
+        SetLastError(RtlNtStatusToDosError(queryStatus));
+        return TIME_ZONE_ID_INVALID;
+    }
+
+    // DST adjustment is disabled
+    if (dayLightEnFlag & 0x2) {
+        return TIME_ZONE_ID_UNKNOWN;
+    }
+
+    // Query all DST info from EEPROM
+    // FIXME One large EEPROM access could be better
+    queryStatus = ExQueryNonVolatileSetting(XC_TZ_DLT_DATE, &type, &dayLightStartDate, sizeof(dayLightStartDate), NULL);
+    if (!NT_SUCCESS(queryStatus)) {
+        SetLastError(RtlNtStatusToDosError(queryStatus));
+        return TIME_ZONE_ID_INVALID;
+    }
+    queryStatus = ExQueryNonVolatileSetting(XC_TZ_STD_DATE, &type, &dayLightEndDate, sizeof(dayLightEndDate), NULL);
+    if (!NT_SUCCESS(queryStatus)) {
+        SetLastError(RtlNtStatusToDosError(queryStatus));
+        return TIME_ZONE_ID_INVALID;
+    }
+    queryStatus = ExQueryNonVolatileSetting(XC_TZ_DLT_BIAS, &type, &dayLightBias, sizeof(dayLightBias), NULL);
+    if (!NT_SUCCESS(queryStatus)) {
+        SetLastError(RtlNtStatusToDosError(queryStatus));
+        return TIME_ZONE_ID_INVALID;
+    }
+    queryStatus = ExQueryNonVolatileSetting(XC_TZ_STD_BIAS, &type, &standardBias, sizeof(standardBias), NULL);
+    if (!NT_SUCCESS(queryStatus)) {
+        SetLastError(RtlNtStatusToDosError(queryStatus));
+        return TIME_ZONE_ID_INVALID;
+    }
+    queryStatus = ExQueryNonVolatileSetting(XC_TZ_DLT_NAME, &type, dayLightName, sizeof(dayLightName), NULL);
+    if (!NT_SUCCESS(queryStatus)) {
+        SetLastError(RtlNtStatusToDosError(queryStatus));
+        return TIME_ZONE_ID_INVALID;
+    }
+    queryStatus = ExQueryNonVolatileSetting(XC_TZ_STD_NAME, &type, standardName, sizeof(standardName), NULL);
+    if (!NT_SUCCESS(queryStatus)) {
+        SetLastError(RtlNtStatusToDosError(queryStatus));
+        return TIME_ZONE_ID_INVALID;
+    }
+
+    if (dayLightEndDate.Month > 12 || dayLightStartDate.Month > 12 ||
+        dayLightEndDate.DayOfWeek > 6 || dayLightStartDate.DayOfWeek > 6 ||
+        dayLightEndDate.Hour > 23 || dayLightStartDate.Hour > 23) {
+        return TIME_ZONE_ID_INVALID;
+    }
+
+    // Xbox stores up to 4 characters here
+    for (UCHAR i = 0; i < 4; i++) {
+        lpTimeZoneInformation->StandardName[i] = standardName[i];
+        lpTimeZoneInformation->DaylightName[i] = dayLightName[i];
+    }
+
+    // There was no DST info
+    if (dayLightEndDate.Month == 0 || dayLightStartDate.Month == 0) {
+        return TIME_ZONE_ID_UNKNOWN;
+    }
+
+    lpTimeZoneInformation->StandardDate.wMonth = dayLightEndDate.Month;
+    lpTimeZoneInformation->StandardDate.wDayOfWeek = dayLightEndDate.DayOfWeek;
+    lpTimeZoneInformation->StandardDate.wDay = dayLightEndDate.Day;
+    lpTimeZoneInformation->StandardDate.wHour = dayLightEndDate.Hour;
```

Comment:
nit: Might want to copy this into something which copies dates, or use the MS type?

Reply:
split into a function